source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,467 @@
+//! SD/MMC Driver Service
+//!
+//! Driver for SD memory cards, SDIO cards and (e)MMC.
+//! This kernel driver will implement the actual protocol
+//! (which commands to send and how to interpret the response),
+//! while the platform driver will implement the device specific part
+//! (how to send and receive the data).
+#![warn(missing_docs)]
+use crate::{
+    comms::{
+        kchannel::{KChannel, KConsumer, KProducer},
+        oneshot::{self, Reusable},
+    },
+    mnemos_alloc::containers::FixedVec,
+    registry::{self, known_uuids, Envelope, KernelHandle, RegisteredDriver},
+    Kernel,
+};
+use core::{convert::Infallible, fmt, time::Duration};
+use uuid::Uuid;
+
+////////////////////////////////////////////////////////////////////////////////
+// Service Definition
+////////////////////////////////////////////////////////////////////////////////
+
+/// [Service](crate::services) definition for SD/MMC protocol drivers.
+pub struct SdmmcService;
+
+impl RegisteredDriver for SdmmcService {
+    type Request = Command;
+    type Response = Response;
+    type Error = Error;
+    type Hello = ();
+    type ConnectError = core::convert::Infallible;
+
+    const UUID: Uuid = known_uuids::kernel::SDMMC;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Message and Error Types
+////////////////////////////////////////////////////////////////////////////////
+
+/// Parameters for building a command to be sent to the card.
+///
+/// The command format must follow the SD specification and is sent on the `CMD` line.
+/// It is 48-bit in length, containing a 6-bit command index and 32-bit argument.
+/// Besides that it includes 7-bit CRC and start, transmission and end bit.
+///
+/// The command structure is the same for SD memory, SDIO and MMC (?),
+/// but the content can be very different.
+/// Therefore the content of the commands is decided here in the kernel service,
+/// while the platform driver has the low-level implementation
+/// for how to physically send the necessary bits to the card.
+pub struct Command {
+    /// The numeric value of the command
+    pub index: u8,
+    /// The argument value for the command
+    pub argument: u32,
+    /// Hardware specific configuration that should be applied
+    pub options: HardwareOptions,
+    /// The type of command
+    pub kind: CommandKind,
+    /// The expected type of the response
+    pub rsp_type: ResponseType,
+    /// Whether the response CRC needs to be checked
+    pub rsp_crc: bool,
+    /// Optional buffer for incoming or outgoing data
+    pub buffer: Option<FixedVec<u8>>,
```

Comment:
Here I'm not sure a `FixedVec` is the right thing to use.
In the blocking implementation of oreboot I'm using `&(mut) [Block]` (where a Block is 512 bytes),
but does not look feasible here, since you will probably run into issues with lifetime when passing this
from kernel to driver.

Reply:
i think fixedvec is _probably_ the correct thing to use for this, although i suppose it could also be a `FixedVec<Block>`?

Reply:
That's a nice idea, I'll experiment with this.